demos/sslecho/main.c

```diff
@@ -0,0 +1,330 @@
+    }
+
+    if (SSL_CTX_use_PrivateKey_file(ctx, "key.pem", SSL_FILETYPE_PEM) <= 0) {
+        ERR_print_errors_fp(stderr);
+        exit(EXIT_FAILURE);
+    }
+}
+
+void configure_client_context(SSL_CTX *ctx) {
+    /*
+     * Configure the client to abort the handshake if certificate verification
+     * fails
+     */
+    SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, NULL);
+    if (!SSL_CTX_load_verify_file(ctx, "cert.pem")) {
+        ERR_print_errors_fp(stderr);
+        exit(EXIT_FAILURE);
+    }
+}
+
+void usage() {
+    printf("Usage: sslecho s\n");
+    printf("       --or--\n");
+    printf("       sslecho c ip\n");
+    printf("       c=client, s=server, ip=dotted ip of server\n");
+    exit(1);
+}
+
+int main(int argc, char **argv) {
+    bool isServer;
+    int result;
+
+    SSL_CTX *ssl_ctx = NULL;
+    SSL *ssl = NULL;
+
+    int server_skt = -1;
+    int client_skt = -1;
+
+    char *txbuf = (char*) OPENSSL_malloc(128);
+    size_t txcap = 128;
+    int txlen;
+
+    char *rxbuf = (char*) OPENSSL_malloc(128);
+    int rxcap = 128;
+    int rxlen;
+
+    char *rem_server_ip = NULL;
+
+    struct sockaddr_in addr;
+    unsigned int addr_len = sizeof(addr);
+
+    /* Splash */
+    printf("\nsslecho : Revision %s : %s : %s\n\n", revision, __DATE__,
+    __TIME__);
+
+    /* Need to know if client or server */
+    if (argc < 2) {
+        usage();
+        /* NOTREACHED */
+    }
+    isServer = (argv[1][0] == 's') ? true : false;
+    /* If client get remote server address (could be 127.0.0.1) */
+    if (!isServer) {
+        if (argc != 3) {
+            usage();
+            /* NOTREACHED */
+        }
+        rem_server_ip = argv[2];
+    }
+
+    /* Create context used by both client and server */
+    ssl_ctx = create_context(isServer);
+
+    /* If server */
+    if (isServer) {
+
+        printf("We are the server on port: %d\n\n", server_port);
+
+        /* Configure server context with appropriate key files */
+        configure_server_context(ssl_ctx);
+
+        /* Create server socket; will bind with server port and listen */
+        server_skt = create_socket(true);
+
+        /*
+         * Loop to accept clients.
+         * Need to implement timeouts on TCP & SSL connect/read functions
+         * before we can catch a CTRL-C and kill the server.
+         */
+        while (server_running) {
+            /* Wait for TCP connection from client */
+            client_skt = accept(server_skt, (struct sockaddr*) &addr,
+                    &addr_len);
+            if (client_skt < 0) {
+                perror("Unable to accept");
+                exit(EXIT_FAILURE);
+            }
+
+            printf("Client TCP connection accepted\n");
+
+            /* Create server SSL structure using newly accepted client socket */
+            ssl = SSL_new(ssl_ctx);
+            SSL_set_fd(ssl, client_skt);
+
+            /* Wait for SSL connection from the client */
+            if (SSL_accept(ssl) <= 0) {
```

Comment:
Here, in the handshake, and in subsequent "application data" reads/writes the code is doing *blocking* I/O, which is not particularly typical.  Far more common are applications with event loops that even for a single connection at least implement timeouts, but otherwise support multiple concurrent connections.

I think that a demo client and server do need to demonstrate an event loop with `SSL_ERROR_WANT_READ`, ...

Reply:
> I think that a demo client and server do need to demonstrate an event loop with SSL_ERROR_WANT_READ, ...

I disagree. I think a demo showing that would be very useful - but it does not need to be *this* demo. In fact I suspect (at least on the client side), blocking IO is quite common for simple apps and I would prefer not to over complicate this demo.

Reply:
Well, perhaps a *really* minimal blocking client is plausibly useful, but surely not the server (unless run as a one-shot process under `inetd`).  And so likely the client and server should be in separate source files, and the *server* needs an event loop, which actually can be used in clients too, the code is basically the same.  See, for example: https://github.com/vdukhovni/postfix/blob/master/postfix/src/tls/tls_bio_ops.c#L152-L294
which supports timed SSL I/O in both the client and server in Postfix.

Reply:
I've actually re-written this demo wherein the server accepts multiple clients. It uses poll() for both server and client connections.
Unfortunately, it is written in C++. It has server and client objects. You could instantiate multiple servers on one machine from a single application.

Reply:
That program was developed to be an API server for a web app. So multiple clients could connect with SSL and exchange data using the protocol of the API.

Reply:
Note that Postfix generally employs a dedicated process for each connection, and the code I linked to employs non-blocking I/O largely to implement a timeout.

This is simple enough, and does not require a multi-client event loop.  Just a timed poll and retry on WANT READ/WRITE.  Blocking forever on network I/O is generally unwise. One might at least set up TCP_KEEPALIVE, but that's a rather blunt tool.